### PR TITLE
Add acceptance tests for how provider handles `request_reason` argument

### PR DESCRIPTION
**Pull request by SarahFrench**
This PR:

- test how the provider behaves when provider configuration arguments come from different sources (config vs ENVs)
- schema-level validation that's in place, e.g. handling of empty strings
    -  See note below
- use cases: how does this argument impact the providers behaviour in plan/apply
    -  This cannot be tested as the `X-Goog-Request-Reason` header is for auditing purposes in GCP

Writing these tests made me realise that empty string validation doesn't cover request_reason, so I opened: https://github.com/hashicorp/terraform-provider-google/issues/19643 . The acceptance tests are written to define the current behaviour and my expectation is that the PF schema will be update in future to add validation, make those tests fail, and the tests will be updated.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

**Comment by SarahFrench** [comment]
Tests running in TeamCity:
- SDK [TestAccSdkProvider_request_reason](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_MMUPSTREAMTESTS_GOOGLE_PACKAGE_PROVIDER/240398?mode=branches&buildTab=overview&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true)
- PF [TestAccFwProvider_request_reason](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_MMUPSTREAMTESTS_GOOGLE_PACKAGE_FWPROVIDER/240397?buildTab=buildParameters&buildParametersTab=parameters&buildParametersShowDependencies=false)

**Review by shuyama1 — APPROVED** [review]

**Review by melinath — APPROVED** [review]

File: mmv1/third_party/terraform/fwprovider/framework_provider_request_reason_test.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+// It is a counterpart to TestAccSdkProvider_request_reason
+func TestAccFwProvider_request_reason(t *testing.T) {
+	testCases := map[string]func(t *testing.T){
+		// Configuring the provider using inputs
+		"config takes precedence over environment variables":                                          TestAccFwProvider_request_reason_configPrecedenceOverEnvironmentVariables,
+		"when request_reason is unset in the config, environment variables are used in a given order": TestAccFwProvider_request_reason_precedenceOrderEnvironmentVariables, // CLOUDSDK_CORE_REQUEST_REASON
+
+		// Schema-level validation
+		// TODO: https://github.com/hashicorp/terraform-provider-google/issues/19643
+		"when request_reason is set to an empty string in the config the value is not ignored, and there isn't any validation about this that raises an error": TestAccFwProvider_request_reason_emptyStringValidation,
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/fwprovider/framework_provider_request_reason_test.go:21]:
```suggestion
		// Configuring the provider using inputs
		"config takes precedence over environment variables":                                          testAccFwProvider_request_reason_configPrecedenceOverEnvironmentVariables,
		"when request_reason is unset in the config, environment variables are used in a given order": testAccFwProvider_request_reason_precedenceOrderEnvironmentVariables, // CLOUDSDK_CORE_REQUEST_REASON

		// Schema-level validation
		// TODO: https://github.com/hashicorp/terraform-provider-google/issues/19643
		"when request_reason is set to an empty string in the config the value is not ignored, and there isn't any validation about this that raises an error": testAccFwProvider_request_reason_emptyStringValidation,
```

File: mmv1/third_party/terraform/fwprovider/framework_provider_request_reason_test.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		// yet, we don't have a race condition
+		// see https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			tc(t)
+		})
+	}
+}
+
+func TestAccFwProvider_request_reason_configPrecedenceOverEnvironmentVariables(t *testing.T) {
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/fwprovider/framework_provider_request_reason_test.go:40]:
```suggestion
func testAccFwProvider_request_reason_configPrecedenceOverEnvironmentVariables(t *testing.T) {
```

File: mmv1/third_party/terraform/fwprovider/framework_provider_request_reason_test.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+				Config: testAccFwProvider_request_reason_inProviderBlock(context),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.google_provider_config_plugin_framework.default", "request_reason", providerReason),
+				),
+			},
+		},
+	})
+}
+
+func TestAccFwProvider_request_reason_precedenceOrderEnvironmentVariables(t *testing.T) {
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/fwprovider/framework_provider_request_reason_test.go:68]:
```suggestion
func testAccFwProvider_request_reason_precedenceOrderEnvironmentVariables(t *testing.T) {
```

File: mmv1/third_party/terraform/fwprovider/framework_provider_request_reason_test.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+				Config: testAccFwProvider_request_reason_inEnvsOnly(context),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr("data.google_provider_config_plugin_framework.default", "request_reason", CLOUDSDK_CORE_REQUEST_REASON),
+				),
+			},
+		},
+	})
+}
+
+func TestAccFwProvider_request_reason_emptyStringValidation(t *testing.T) {
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/fwprovider/framework_provider_request_reason_test.go:97]:
```suggestion
func testAccFwProvider_request_reason_emptyStringValidation(t *testing.T) {
```

**Review by NickElliot — APPROVED** [review]
took a moment to follow how the tests were operating, but LGTM!

**Comment by SarahFrench** [comment]
Thanks for the review! Yeah, I probably should have left a note about the 2 data sources used in these acceptance tests. You've read through it now but they basically show what values the SDK and PF sides of the provider have been configured with. Basically non-user facing versions of [google_client_config](https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/client_config) that expose all values impacted by user-supplied arguments. It makes tests a lot easier versus trying to test things indirectly, though indirect testing is still needed for the 'usage' test cases in particular.

**Pull request merged**